javascript/node-oauth-demo/index.js

```diff
@@ -1,148 +1,147 @@
+ * never exposed to client-side JavaScript (via cookies or localStorage),
+ * and refresh tokens should be kept httpOnly and encrypted if possible.
+ * 
+ * For official Asana docs, see: https://developers.asana.com/docs/oauth
+ */
+
 require("dotenv").config();
-const axios = require("axios");
 const express = require("express");
+const axios = require("axios");
 const path = require("path");
 const cors = require("cors");
 const cookieParser = require("cookie-parser");
 const { v4: uuidv4 } = require("uuid");
 
 const app = express();
+const PORT = process.env.PORT || 3000;
 
-// Enable CORS and assume the allowed origin is the redirect URI.
-// i.e., this assumes that your client shares the same domain as the server.
-app.use(
-  cors({
-    origin: "http://localhost:3000",
-  })
-);
+// Enable CORS for local frontend
+app.use(cors({ origin: "http://localhost:3000" }));
 
-// Enable storage of data in cookies.
-// Signed cookies are signed by the COOKIE-SECRET environment variable.
+// Parse cookies, including signed cookies (used to validate state during OAuth flow)
 app.use(cookieParser(process.env.COOKIE_SECRET));
 
-// Serve files in the ./static folder.
+// Serve static HTML and assets from the "static" directory
 app.use(express.static("static"));
 
-// Send static index.html page to the client.
-// This page includes a button to authenticatate with Asana.
+// Serve landing page with "Authenticate with Asana" button
 app.get("/", (req, res) => {
   res.sendFile(path.join(__dirname, "/static/index.html"));
 });
 
-// When the user clicks on the "Authenticate with Asana" button (from index.html),
-// it redirects them to the user authorization endpoint.
-// Docs: https://developers.asana.com/docs/oauth#user-authorization-endpoint
+/**
+ * This route is triggered when the user clicks "Authenticate with Asana".
+ * It generates a CSRF-protection `state`, stores it in a signed cookie,
+ * and redirects the user to Asana's authorization URL to begin the OAuth flow.
+ *
+ * Docs: https://developers.asana.com/docs/oauth#user-authorization-endpoint
+ */
 app.get("/authenticate", (req, res) => {
-  // Generate a `state` value and store it
-  // Docs: https://developers.asana.com/docs/oauth#response
-  let generatedState = uuidv4();
+  const state = uuidv4();
 
-  // Expiration of 5 minutes
-  res.cookie("state", generatedState, {
-    maxAge: 1000 * 60 * 5,
+  // Store state in a signed cookie to protect against CSRF.
+  res.cookie("state", state, {
+    maxAge: 5 * 60 * 1000, // 5 minutes
     signed: true,
+    sameSite: "lax",
   });
 
-  res.redirect(
-    `https://app.asana.com/-/oauth_authorize?response_type=code&client_id=${process.env.CLIENT_ID}&redirect_uri=${process.env.REDIRECT_URI}&state=${generatedState}`
-  );
+  const authUrl = `https://app.asana.com/-/oauth_authorize?response_type=code&client_id=${process.env.CLIENT_ID}&redirect_uri=${process.env.REDIRECT_URI}&state=${state}`;
```

Comment:
[Nit]: Maybe we hard code the `redirect_uri=${process.env.REDIRECT_URI}` since we end up coming back to the `/oauth-callback` endpoint below.

Essentially in the `README.md` we tell users to add a `REDIRECT_URI` variable to the `.env` file and also add it to Asana allow redirect_uri list `Take note of your client ID and client secret, and set the redirect URI to http://localhost:3000/oauth-callback.`. We can save the user a step by just hard coding the `redirect_uri` it also keeps the code less confusing